File: tests/basic/arith_simplify.ml
open Optitrust
open Target

(* TODO: the operations don't seem to work here; an issue with the target? *)

let _ = Run.doc_script_cpp (fun _ ->
    !! Arith_basic.(simpl gather) [cVarDef "a"; dBody];
       Arith_basic.(simpl gather) [cVarDef "b"; dBody];
       Arith_basic.(simpl gather) [cVarDef "c"; dBody];
       Arith_basic.(simpl expand) [cVarDef "d"; dBody];
  )
"
int main() {
  int a = 2 + 3;
  int b = 3 * a + 4 * a;
  int c = (a * 3 * a) / a;
  int d = (a + b) * c;
}
"



(* LATER: add a "compute" transformation to simplify
    - products of int
    - sums and products of doubles

   LATER: simplification recursively in atoms, see example of [w];
   to implement using trm_map. *)


let _ = Run.script_cpp (fun _ ->

    !! Arith_basic.(simpl identity) [nbMulti; cWriteVar "x"; dRHS];
    !! Arith_basic.(simpl normalize) [nbMulti; cWriteVar "x"; dRHS];
    !! Arith_basic.(simpl gather) [nbMulti; cWriteVar "y"; dRHS];
    !! Arith_basic.(simpl gather) [nbMulti; cWriteVar "z"; dRHS];
    !! Arith_basic.(simpl gather_rec) [nbMulti; cWriteVar "t"; dRHS];
    !! Arith_basic.(simpl expand) [nbMulti; cWriteVar "u"; dRHS];
    !! Arith_basic.(simpl expand) [nbMulti; cWriteVar "v"; dRHS];

(* For testing one line, add a line in the source "r = ...;" and use:
    !! Arith_basic.(simpl gather_rec) [nbMulti; cWriteVar "r"; dRHS];
*)

(* For testing all lines concerning one variable:
    !! (for i = 0 to 5 do
         Printf.printf "===%d====\n" i;
         Arith_basic.(simpl gather_rec) [occIndex i; cWriteVar "t"; dRHS];
    done);
*)

(* For debugging apply_bottom_up:
     Arith_basic.(simpl (apply_bottom_up_if_debug true true)) [occIndex i; cWriteVar "t"; dRHS];
     Arith_basic.(simpl apply_bottom_up_debug) [occIndex i; cWriteVar "t"; dRHS];
*)


)

File: tests/basic/arith_simplify.cpp
int f(int x) {
  return x;
}

int main()
{
  double a, b, c, d, e;
  double r, s, t, u, v, w, x, y, z;

  // test parsing and normalizing
  x = 2 * b;
  x = a + b * 2 - (3 * c) - d;
  x = a + ((a + 2*b) + 1);
  x = 3 + 5.0 + a + (a + b*2 - 3*c);
  x = 3 * 5.0 * a * b * b;
  x = a / (b * b* c) / d;

  // test gather sum
  y = 1 + 2; // = 3
  y = 1 + 3.2 + 4 + 5.2; // = 5 + 3.2 + 5.2
  y = a + a + a - a; // = 2*a
  y = 2*a - 3*a + b - 5*b; // = - a -4*b
  y = 3*a + 2*b - b + c - 4*a ; // = -a + b + c

  // test gather prod
  z = a * a * a / a; // = a * a
  z = a / b / a * b / b; // = 1 / b

  // test gather nested
  t = (a + (3 + 2)*a); // = 6 * a
  t = a + b * a / b; // = 2 * a
  t = (a + (3 + 2)*a) / a * b; // = 6 * b
  t = a / (b * a * b) / (1 / b); // = 1 / b

  // test expand
  u = a * (b + c); // = a * b + a * c
  u = (a + b) * (c + d); // = a * c + a * d + b * c + b * d
  u = 5*a * (2*b + 3*c + 4*d);

  // test expand from pic demo
  v = (a / b * c / (a * d * d / b * e)); // = c / (d * d) / e
  v = (a * b * c / a); // = b * c
  v = ((a / (b / c)) + d * c / b) * (b / c); // = a + d
  v = ((a * ((b / c) * (d / ((((b * a) * a) / c) / e)))) * (a / e)); // = d
  v = (((a * d) / (a / e)) * (1. / e)); // = d
  v = (((a / (1. / e)) + (b * (c / (b / e )))) * (1. / e)); // = a + c


  // test for recursion in atoms (LATER)
  w = a + a * 3 + f(b * f(c + c) / b); // = 4*a + f(f(2*c))

  return 0;
}

